pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,208 @@ package heal
 }
+
+func (u *healClient) listenToConnectionChanges(heal healRequestFuncType, currentPathSegment *networkservice.PathSegment, errCh chan error) {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{
+		PathSegments: []*networkservice.PathSegment{{Name: currentPathSegment.Name}, {Name: ""}},
+	})
+	if err != nil {
+		errCh <- errors.Wrap(err, "MonitorConnections failed")
+		return
+	}
+	close(errCh)
+
+	for {
+		event, err := monitorClient.Recv()
+		if err != nil {
+			if heal != nil {
+				<-u.conCacheExecutor.AsyncExec(func() {
+					for id := range u.connCache {
+						heal(id, true)
+					}
+				})
+			}
+			return
+		}
+
+		switch event.GetType() {
+		case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER:
+			continue
+		case networkservice.ConnectionEventType_UPDATE:
+			<-u.conCacheExecutor.AsyncExec(func() {
+				for _, conn := range event.GetConnections() {
+					if _, ok := u.connCache[conn.GetId()]; ok {
+						u.connCache[conn.GetId()] = conn.Clone()
+					}
+				}
+			})
+
+			u.checkExpectedConnections(event.GetConnections())
+
+		case networkservice.ConnectionEventType_DELETE:
+			if heal != nil {
+				var connsToHeal []string
+				<-u.conCacheExecutor.AsyncExec(func() {
+					for _, eventConn := range event.GetConnections() {
+						if conn, ok := u.connCache[eventConn.GetPrevPathSegment().GetId()]; ok {
+							connsToHeal = append(connsToHeal, conn.GetId())
+						}
+					}
+				})
+				for _, id := range connsToHeal {
+					heal(id, false)
+				}
+			}
+		}
+	}
+}
+
+func (u *healClient) checkExpectedConnections(eventConns map[string]*networkservice.Connection) {
+	u.expectedConns.Range(func(reqId string, reqChs connVerificationRequest) bool {
+		for _, conn := range eventConns {
+			if reqId == conn.GetPrevPathSegment().GetId() {
+				select {
+				case <-reqChs.stopCh:
+					u.expectedConns.Delete(reqId)
+					return false
+				case reqChs.connectionSuccessCh <- conn.GetCurrentPathSegment():
+					break
+				}
+			}
+		}
+		return true
+	})
+}
+
+func (u *healClient) addToExpectedConnections(conn *networkservice.Connection) *connVerificationRequest {
+	connAlreadyExists := false
+	<-u.conCacheExecutor.AsyncExec(func() {
+		if cachedConn, ok := u.connCache[conn.GetId()]; ok {
+			if cachedConn.GetCurrentPathSegment().GetId() == conn.GetCurrentPathSegment().GetId() {
+				connAlreadyExists = true
+			}
+		}
+	})
+
+	if connAlreadyExists {
+		return nil
+	}
+
+	req := connVerificationRequest{
+		connectionSuccessCh: make(chan *networkservice.PathSegment, 10),
+		stopCh:              make(chan struct{}),
+	}
+
+	u.expectedConns.Store(conn.GetId(), req)
+
+	return &req
+}
+
+func (u *healClient) confirmConnectionSuccess(conn *networkservice.Connection, req *connVerificationRequest) error {
+	timeoutCh := time.After(time.Millisecond * 100)
+	for {
+		select {
+		case segment := <-req.connectionSuccessCh:
+			if conn.GetNextPathSegment().Equal(segment) {
```

Comment:
I don't think that we need such check. We have already filtered out all connections except the ones going from us. We now that for the `segment` previous `conn.Id` matches this `conn.Id`. So there is no need to send a `segment` and check it for something, because we have already checked everything we want.